### ISSUE TITLE
Several lint auto-fix produce non-compilable code

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-clippy#10608]
### Summary

Running example code with `--fix` produces a broken code

### Lint Name

redundant_slicing AND useless_asref

### Reproducer

I tried this code:

```rust
fn main() {
    let v_ref: &[i32] = &[1, 2, 3];
    let v_mut: &mut [i32] = &mut [0, 0, 0];
    call_with_fn_mut(|x| {
        let values = if x > 0 { v_ref } else { &v_mut[..] }; // SAME ERROR with   v_mut.as_ref()
        println!("{x} = {values:?}");
    });
}

fn call_with_fn_mut<F: FnMut(i32)>(mut _callback: F) {}
```

I saw this happen:

```
warning: redundant slicing of the whole range
 --> src/main.rs:5:48
  |
5 |         let values = if x > 0 { v_ref } else { &v_mut[..] };
  |                                                ^^^^^^^^^^ help: use the original value instead: `v_mut`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_slicing
  = note: `#[warn(clippy::redundant_slicing)]` on by default
```

I expected not to see a warning, because if I follow the advice, its a compiler error:

```rust
error[E0507]: cannot move out of `v_mut`, a captured variable in an `FnMut` closure
 --> src/main.rs:5:48
  |
3 |     let v_mut: &mut [i32] = &mut [0, 0, 0];
  |         ----- captured outer variable
4 |     call_with_fn_mut(|x| {
  |                      --- captured by this `FnMut` closure
5 |         let values = if x > 0 { v_ref } else { v_mut };
  |                                                ^^^^^ move occurs because `v_mut` has type `&mut [i32]`, which does not implement the `Copy` trait
```

### Version

```text
rustc 1.68.2 (9eb3afe9e 2023-03-27)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: x86_64-unknown-linux-gnu
release: 1.68.2
LLVM version: 15.0.6
```

**Comment by llogiq** [comment]
Do I understand correctly that it appears the lint ignores the difference in mutability?

**Comment by nyurik** [comment]
I'm not certain of the course, but it does seem plausable

**Comment by rsammelson** [comment]
I think this might be another instance of the same bug:

```rs
struct Test<'l> {
    data: &'l mut [u8],
}

impl Test<'_> {
    fn test(&self) -> impl Iterator<Item = &[u8]> {
        // Removing as_ref gives: return type was inferred to be `std::iter::Once<&mut [u8]>` here
        core::iter::once(self.data.as_ref()) // warning: this call to `as_ref` does nothing
    }
}
```
